src/classification/mondrian_tree.rs

```diff
@@ -0,0 +1,373 @@
+        let has_duplicates = children.iter().any(|item| !seen.insert(item));
+        assert!(
+            !has_duplicates,
+            "Multiple nodes share one child. Children left: {:?}, Children right: {:?}",
+            children_l, children_r
+        );
+    }
+
+    fn compute_split_time(
+        &self,
+        time: F,
+        exp_sample: F,
+        node_idx: usize,
+        y: usize,
+        extensions_sum: F,
+    ) -> F {
+        if self.nodes[node_idx].is_dirac(y) {
+            // println!(
+            //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - all same class",
+            //     extensions_sum
+            // );
+            return F::zero();
+        }
+
+        if extensions_sum > F::zero() {
+            let split_time = time + exp_sample;
+
+            // From River: If the node is a leaf we must split it
+            if self.nodes[node_idx].is_leaf {
+                // println!(
+                //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - split is_leaf",
+                //     extensions_sum
+                // );
+                return split_time;
+            }
+
+            // From River: Otherwise we apply Mondrian process dark magic :)
+            // 1. We get the creation time of the childs (left and right is the same)
+            let child_idx = self.nodes[node_idx].left.unwrap();
+            let child_time = self.nodes[child_idx].time;
+            // 2. We check if splitting time occurs before child creation time
+            if split_time < child_time {
+                // println!(
+                //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - split mid tree",
+                //     extensions_sum
+                // );
+                return split_time;
+            }
+            // println!("go_downwards() - node: {node_idx} - extensions_sum: {:?} - not increased enough to split (mid node)", extensions_sum);
+        } else {
+            // println!(
+            //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - not outside box",
+            //     extensions_sum
+            // );
+        }
+
+        F::zero()
+    }
+
+    fn go_downwards(&mut self, node_idx: usize, x: &Array1<F>, y: usize) -> usize {
+        let time = self.nodes[node_idx].time;
+        let node_min_list = &self.nodes[node_idx].min_list;
+        let node_max_list = &self.nodes[node_idx].max_list;
+        let extensions = {
+            let e_min = (node_min_list - x).mapv(|v| F::max(v, F::zero()));
+            let e_max = (x - node_max_list).mapv(|v| F::max(v, F::zero()));
+            &e_min + &e_max
+        };
+        // 'T' in River
+        let exp_sample = {
+            let lambda = extensions.sum();
+            let exp_dist = Exp::new(lambda.to_f32().unwrap()).unwrap();
+            let exp_sample = F::from_f32(exp_dist.sample(&mut self.rng)).unwrap();
+            // DEBUG: shadowing with Exp expected value
```

Comment:
I am performing multiple waves of reading and comparing your code against the reference repository you mentioned in the PR, River, and one_learn. 

I see you left a comment here., so there might be a reason. Why are you not using `1 / exp`, like in the reference code?

Reply:
About the input parameter to the exponential distribution Rust's [rand_distr::Exp](https://docs.rs/rand_distr/latest/rand_distr/struct.Exp.html#method.new) takes directly lambda, while numpy's [np.random.exponential](https://numpy.org/doc/stable/reference/random/generated/numpy.random.exponential.html) takes 1/lambda.

I later shadowed exp_sample to remove stochasticity and have consistent results between codebases. The results are the same with or without it.